reward/staking_info.go

```diff
@@ -174,6 +200,77 @@ func (s *StakingInfo) DecodeRLP(st *rlp.Stream) error {
 	return nil
 }
 
+func (s *StakingInfo) GetConsolidatedStakingInfo() *ConsolidatedStakingInfo {
+	c := &ConsolidatedStakingInfo{
+		nodes:     make([]consolidatedNode, 0),
+		nodeIndex: make(map[common.Address]int),
+	}
+
+	rewardIndex := make(map[common.Address]int) // temporarily map rewardAddr -> index in []nodes
+
+	for j := 0; j < len(s.CouncilNodeAddrs); j++ {
+		var (
+			nodeAddr      = s.CouncilNodeAddrs[j]
+			stakingAddr   = s.CouncilStakingAddrs[j]
+			rewardAddr    = s.CouncilRewardAddrs[j]
+			stakingAmount = s.CouncilStakingAmounts[j]
+		)
+		if idx, ok := rewardIndex[rewardAddr]; !ok {
+			c.nodes = append(c.nodes, consolidatedNode{
+				NodeAddrs:     []common.Address{nodeAddr},
+				StakingAddrs:  []common.Address{stakingAddr},
+				RewardAddr:    rewardAddr,
+				StakingAmount: stakingAmount,
+			})
+			c.nodeIndex[nodeAddr] = len(c.nodes) - 1 // point to new element
+			rewardIndex[rewardAddr] = len(c.nodes) - 1
+		} else {
+			c.nodes[idx].NodeAddrs = append(c.nodes[idx].NodeAddrs, nodeAddr)
+			c.nodes[idx].StakingAddrs = append(c.nodes[idx].StakingAddrs, stakingAddr)
+			c.nodes[idx].StakingAmount += stakingAmount
+			c.nodeIndex[nodeAddr] = idx // point to existing element
+		}
+	}
+	return c
+}
+
+func (c *ConsolidatedStakingInfo) GetAllNodes() []consolidatedNode {
+	return c.nodes
+}
+
+func (c *ConsolidatedStakingInfo) GetConsolidatedNode(nodeAddr common.Address) *consolidatedNode {
+	if idx, ok := c.nodeIndex[nodeAddr]; ok {
+		return &c.nodes[idx]
+	} else {
+		return nil
+	}
+}
+
+// Calculate Gini coefficient of the StakingAmounts.
+// Only amounts greater or equal to `minStake` are included in the calculation.
+// Set `minStake` to 0 to calculate Gini coefficient of all amounts.
+func (c *ConsolidatedStakingInfo) CalcGiniCoefficient(minStake uint64) float64 {
```

Comment:
We have the cases where all nodes doesn't have enough amount of staking. Do you think we should take consider the case?

Reply:
Currently it returns `-1`, right? Both makes sense to me, but if `NaN` makes inefficient then I think the original change seems better to me. I resolve this comment, but you may be able to update this.

Reply:
- I meant to say `NaN` makes it more efficient. (by distinguishing "never calculated" and "calculated but empty array" in fillMissingGiniCoefficient).
- But after some research I found that having `NaN` in JSON is problematic in [Java](https://stackoverflow.com/questions/15228651/how-to-parse-json-string-containing-nan-in-node-js). So let's not use `NaN` anyway.